app/models/foreman_chef/fact_importer.rb

```diff
@@ -30,9 +32,21 @@ def add_missing_facts(tree_hash, parent = nil, prefix = '')
         if fact_names[name_with_prefix].present?
           fact_name_id = fact_names[name_with_prefix]
         else
-          fact_name_id = fact_name_class.create!(:name      => name_with_prefix,
-                                                 :parent_id => parent,
-                                                 :compose   => compose).id
+          # fact names didn't contain fact_name so we create new one
+          fact_name = fact_name_class.new(:name      => name_with_prefix,
+                                          :parent_id => parent,
+                                          :compose   => compose)
+          if fact_name.save
+            fact_name_id = fact_name.id
+          elsif fact_name.errors[:name].present?
+            # saving could have failed because of raise condition with another import process,
+            # so if the error is on name uniqueness, we try to reload conflicting name and use it
+            conflicting_fact_name = fact_name_class.find_by_name(fact_name.name)
```

Comment:
please use where(:name => fact_name.name).first as a preparation for the rails 4 conventions (the find_by_field methods use method_missing and perform very poorly, they're deprecated)